#! /bin/sh -e

# gdc-4.1.dpatch by Arthur Loiret <arthur.loiret@gmail.com>
# DP: Setup gcc-4.1.x for D

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


diff -ruN gcc/DEV-PHASE gcc/DEV-PHASE
--- gcc/DEV-PHASE	2007-10-22 06:54:31.696246014 +0000
+++ gcc/DEV-PHASE	2007-10-22 07:20:13.288113604 +0000
@@ -1 +1 @@
-prerelease
+prerelease gdc 0.25 20080419, using dmd 1.024
diff -ruN gcc/cgraph.c gcc/cgraph.c
--- gcc/cgraph.c	2007-10-22 06:54:19.984995402 +0000
+++ gcc/cgraph.c	2007-10-22 07:20:12.859994795 +0000
@@ -182,6 +182,7 @@
 cgraph_node (tree decl)
 {
   struct cgraph_node key, *node, **slot;
+  tree context;
 
   gcc_assert (TREE_CODE (decl) == FUNCTION_DECL);
 
@@ -203,12 +204,16 @@
   node = cgraph_create_node ();
   node->decl = decl;
   *slot = node;
-  if (DECL_CONTEXT (decl) && TREE_CODE (DECL_CONTEXT (decl)) == FUNCTION_DECL)
+  if (!DECL_NO_STATIC_CHAIN (decl))
     {
-      node->origin = cgraph_node (DECL_CONTEXT (decl));
-      node->next_nested = node->origin->nested;
-      node->origin->nested = node;
-      node->master_clone = node;
+      context = decl_function_context (decl);
+      if (context)
+        {
+	  node->origin = cgraph_node (context);
+	  node->next_nested = node->origin->nested;
+	  node->origin->nested = node;
+	  node->master_clone = node;
+        }
     }
   return node;
 }
diff -ruN gcc/config/arm/arm.c gcc/config/arm/arm.c
--- gcc/config/arm/arm.c	2007-10-22 06:55:10.406990680 +0000
+++ gcc/config/arm/arm.c	2007-10-22 07:20:12.900005899 +0000
@@ -15382,6 +15382,15 @@
 	  asm_fprintf (asm_out_file, "\t.movsp %r, #%d\n",
 		       REGNO (e0), (int)INTVAL(XEXP (e1, 1)));
 	}
+      else if (GET_CODE (e1) == PLUS
+	      && GET_CODE (XEXP (e1, 0)) == REG
+	      && REGNO (XEXP (e1, 0)) == SP_REGNUM
+	      && GET_CODE (XEXP (e1, 1)) == CONST_INT)
+	{
+	  /* Set reg to offset from sp.  */
+	  asm_fprintf (asm_out_file, "\t.movsp %r, #%d\n",
+		       REGNO (e0), (int)INTVAL(XEXP (e1, 1)));
+	}
       else
 	abort ();
       break;
diff -ruN gcc/config/darwin.h gcc/config/darwin.h
--- gcc/config/darwin.h	2007-10-22 06:55:10.194931820 +0000
+++ gcc/config/darwin.h	2007-10-22 07:20:12.908008120 +0000
@@ -926,8 +926,8 @@
 
 #define MACHO_DYNAMIC_NO_PIC_P	(TARGET_DYNAMIC_NO_PIC)
 #define MACHOPIC_INDIRECT	(flag_pic || MACHO_DYNAMIC_NO_PIC_P)
-#define MACHOPIC_JUST_INDIRECT	(flag_pic == 1 || MACHO_DYNAMIC_NO_PIC_P)
-#define MACHOPIC_PURE		(flag_pic == 2 && ! MACHO_DYNAMIC_NO_PIC_P)
+#define MACHOPIC_JUST_INDIRECT	(MACHO_DYNAMIC_NO_PIC_P)
+#define MACHOPIC_PURE		(flag_pic && ! MACHO_DYNAMIC_NO_PIC_P)
 
 #undef TARGET_ENCODE_SECTION_INFO
 #define TARGET_ENCODE_SECTION_INFO  darwin_encode_section_info
diff -ruN gcc/config/i386/i386.c gcc/config/i386/i386.c
--- gcc/config/i386/i386.c	2007-10-22 06:55:07.418161092 +0000
+++ gcc/config/i386/i386.c	2007-10-22 07:20:12.936015892 +0000
@@ -5018,6 +5018,11 @@
     frame->red_zone_size = 0;
   frame->to_allocate -= frame->red_zone_size;
   frame->stack_pointer_offset -= frame->red_zone_size;
+
+  if (cfun->naked)
+      /* As above, skip return address */
+      frame->stack_pointer_offset = UNITS_PER_WORD;
+
 #if 0
   fprintf (stderr, "nregs: %i\n", frame->nregs);
   fprintf (stderr, "size: %i\n", size);
@@ -17346,7 +17351,7 @@
 	  output_set_got (tmp);
 
 	  xops[1] = tmp;
-	  output_asm_insn ("mov{l}\t{%0@GOT(%1), %1|%1, %0@GOT[%1]}", xops);
+	  output_asm_insn ("mov{l}\t{%a0@GOT(%1), %1|%1, %a0@GOT[%1]}", xops);
 	  output_asm_insn ("jmp\t{*}%1", xops);
 	}
     }
diff -ruN gcc/config/rs6000/rs6000.c gcc/config/rs6000/rs6000.c
--- gcc/config/rs6000/rs6000.c	2007-10-22 06:55:09.006601984 +0000
+++ gcc/config/rs6000/rs6000.c	2007-10-22 07:20:12.968024775 +0000
@@ -15306,7 +15306,8 @@
 	 use language_string.
 	 C is 0.  Fortran is 1.  Pascal is 2.  Ada is 3.  C++ is 9.
 	 Java is 13.  Objective-C is 14.  */
-      if (! strcmp (language_string, "GNU C"))
+      if (! strcmp (language_string, "GNU C") ||
+ 	  ! strcmp (language_string, "GNU D"))
 	i = 0;
       else if (! strcmp (language_string, "GNU F77")
 	       || ! strcmp (language_string, "GNU F95"))
diff -ruN gcc/dwarf2out.c gcc/dwarf2out.c
--- gcc/dwarf2out.c	2007-10-22 06:54:38.122029574 +0000
+++ gcc/dwarf2out.c	2007-10-22 07:20:12.996032548 +0000
@@ -5322,7 +5322,8 @@
   unsigned int lang = get_AT_unsigned (comp_unit_die, DW_AT_language);
 
   return (lang == DW_LANG_C || lang == DW_LANG_C89
-	  || lang == DW_LANG_C_plus_plus);
+	  || lang == DW_LANG_C_plus_plus
+	  || lang == DW_LANG_D);
 }
 
 /* Return TRUE if the language is C++.  */
@@ -12213,6 +12214,8 @@
     language = DW_LANG_Pascal83;
   else if (strcmp (language_string, "GNU Java") == 0)
     language = DW_LANG_Java;
+  else if (strcmp (language_string, "GNU D") == 0)
+    language = DW_LANG_D;
   else
     language = DW_LANG_C89;
 
@@ -13355,7 +13358,7 @@
 
       /* For local statics lookup proper context die.  */
       if (TREE_STATIC (decl) && decl_function_context (decl))
-	context_die = lookup_decl_die (DECL_CONTEXT (decl));
+	context_die = lookup_decl_die (decl_function_context (decl));
 
       /* If we are in terse mode, don't generate any DIEs to represent any
 	 variable declarations or definitions.  */
diff -ruN gcc/expr.c gcc/expr.c
--- gcc/expr.c	2007-10-22 06:54:38.122029574 +0000
+++ gcc/expr.c	2007-10-22 07:20:13.016038100 +0000
@@ -8477,6 +8477,11 @@
       /* Lowered by gimplify.c.  */
       gcc_unreachable ();
 
+    case STATIC_CHAIN_EXPR:
+    case STATIC_CHAIN_DECL:
+      /* Lowered by tree-nested.c */
+      gcc_unreachable ();
+
     case EXC_PTR_EXPR:
       return get_exception_pointer (cfun);
 
diff -ruN gcc/function.c gcc/function.c
--- gcc/function.c	2007-10-22 06:54:39.682462693 +0000
+++ gcc/function.c	2007-10-22 07:20:13.032042541 +0000
@@ -2997,7 +2997,8 @@
       FUNCTION_ARG_ADVANCE (all.args_so_far, data.promoted_mode,
 			    data.passed_type, data.named_arg);
 
-      assign_parm_adjust_stack_rtl (&data);
+      if (!cfun->naked)
+	assign_parm_adjust_stack_rtl (&data);
 
       if (assign_parm_setup_block_p (&data))
 	assign_parm_setup_block (&all, parm, &data);
@@ -3012,7 +3013,8 @@
 
   /* Output all parameter conversion instructions (possibly including calls)
      now that all parameters have been copied out of hard registers.  */
-  emit_insn (all.conversion_insns);
+  if (!cfun->naked)
+    emit_insn (all.conversion_insns);
 
   /* If we are receiving a struct value address as the first argument, set up
      the RTL for the function result. As this might require code to convert
@@ -3142,6 +3144,9 @@
   struct assign_parm_data_all all;
   tree fnargs, parm, stmts = NULL;
 
+  if (cfun->naked)
+    return NULL;
+
   assign_parms_initialize_all (&all);
   fnargs = assign_parms_augmented_arg_list (&all);
 
@@ -4176,11 +4181,15 @@
       tree parm = cfun->static_chain_decl;
       rtx local = gen_reg_rtx (Pmode);
 
-      set_decl_incoming_rtl (parm, static_chain_incoming_rtx);
       SET_DECL_RTL (parm, local);
       mark_reg_pointer (local, TYPE_ALIGN (TREE_TYPE (TREE_TYPE (parm))));
 
-      emit_move_insn (local, static_chain_incoming_rtx);
+      if (! cfun->custom_static_chain)
+        {
+	    set_decl_incoming_rtl (parm, static_chain_incoming_rtx);
+	    emit_move_insn (local, static_chain_incoming_rtx);
+	}
+      /* else, the static chain will be set in the main body */
     }
 
   /* If the function receives a non-local goto, then store the
@@ -5093,6 +5102,9 @@
 #endif
   edge_iterator ei;
 
+  if (cfun->naked)
+      return;
+
 #ifdef HAVE_prologue
   if (HAVE_prologue)
     {
diff -ruN gcc/function.h gcc/function.h
--- gcc/function.h	2007-10-22 06:54:39.690464914 +0000
+++ gcc/function.h	2007-10-22 07:20:13.040044762 +0000
@@ -461,6 +461,14 @@
   /* Number of units of floating point registers that need saving in stdarg
      function.  */
   unsigned int va_list_fpr_size : 8;
+
+  /* Nonzero if static chain is initialized by something other than
+     static_chain_incoming_rtx. */
+  unsigned int custom_static_chain : 1;
+
+  /* Nonzero if no code should be generated for prologues, copying
+     parameters, etc. */
+  unsigned int naked : 1;
 };
 
 /* If va_list_[gf]pr_size is set to this, it means we don't know how
diff -ruN gcc/gcc.c gcc/gcc.c
--- gcc/gcc.c	2007-10-22 06:54:20.021005397 +0000
+++ gcc/gcc.c	2007-10-22 07:20:13.092059197 +0000
@@ -126,6 +126,9 @@
 /* Flag set by cppspec.c to 1.  */
 int is_cpp_driver;
 
+/* Flag set by drivers needing Pthreads. */
+int need_pthreads;
+
 /* Flag saying to pass the greatest exit code returned by a sub-process
    to the calling program.  */
 static int pass_exit_codes;
@@ -461,6 +464,7 @@
 	assembler has done its job.
  %D	Dump out a -L option for each directory in startfile_prefixes.
 	If multilib_dir is set, extra entries are generated with it affixed.
+ %N     Output the currently selected multilib directory name.
  %l     process LINK_SPEC as a spec.
  %L     process LIB_SPEC as a spec.
  %G     process LIBGCC_SPEC as a spec.
@@ -879,6 +883,8 @@
 #endif
 #endif
 
+#define GCC_SPEC_FORMAT_4 1
+
 /* Record the mapping from file suffixes for compilation specs.  */
 
 struct compiler
@@ -3818,6 +3824,9 @@
 	}
     }
 
+  if (need_pthreads)
+      n_switches++;
+
   if (save_temps_flag && use_pipes)
     {
       /* -save-temps overrides -pipe, so that temp files are produced */
@@ -4156,6 +4165,18 @@
 	}
     }
 
+  if (need_pthreads)
+    {
+	switches[n_switches].part1 = "pthread";
+	switches[n_switches].args = 0;
+	switches[n_switches].live_cond = SWITCH_OK;
+	/* Do not print an error if there is not expansion for -pthread. */
+	switches[n_switches].validated = 1;
+	switches[n_switches].ordering = 0;
+
+	n_switches++;
+    }
+
   switches[n_switches].part1 = 0;
   infiles[n_infiles].name = 0;
 }
@@ -5136,6 +5157,17 @@
 	      return value;
 	    break;
 
+	  case 'N':
+	    if (multilib_dir)
+	      {
+		arg_going = 1;
+		obstack_grow (&obstack, "-fmultilib-dir=",
+			      strlen ("-fmultilib-dir="));
+	        obstack_grow (&obstack, multilib_dir,
+			      strlen (multilib_dir));
+	      }
+	    break;
+
 	    /* Here we define characters other than letters and digits.  */
 
 	  case '{':
diff -ruN gcc/gcc.h gcc/gcc.h
--- gcc/gcc.h	2007-10-22 06:54:20.021005397 +0000
+++ gcc/gcc.h	2007-10-22 07:20:13.108063638 +0000
@@ -38,7 +38,7 @@
    || (CHAR) == 'e' || (CHAR) == 'T' || (CHAR) == 'u' \
    || (CHAR) == 'I' || (CHAR) == 'm' || (CHAR) == 'x' \
    || (CHAR) == 'L' || (CHAR) == 'A' || (CHAR) == 'V' \
-   || (CHAR) == 'B' || (CHAR) == 'b')
+   || (CHAR) == 'B' || (CHAR) == 'b' || (CHAR) == 'J')
 
 /* This defines which multi-letter switches take arguments.  */
 
diff -ruN gcc/gimplify.c gcc/gimplify.c
--- gcc/gimplify.c	2007-10-22 06:54:52.666066460 +0000
+++ gcc/gimplify.c	2007-10-22 07:20:13.116065859 +0000
@@ -4200,6 +4200,12 @@
 	    }
 	  break;
 
+	case STATIC_CHAIN_EXPR:
+	  /* The argument is used as information only.  No need to gimplify */
+	case STATIC_CHAIN_DECL:  
+	  ret = GS_ALL_DONE;
+	  break;
+	  
 	case TREE_LIST:
 	  gcc_unreachable ();
 
diff -ruN gcc/predict.c gcc/predict.c
--- gcc/predict.c	2007-10-22 06:54:38.342090655 +0000
+++ gcc/predict.c	2007-10-22 07:20:13.128069190 +0000
@@ -1339,6 +1339,7 @@
 	     care for error returns and other cases are often used for
 	     fast paths trought function.  */
 	  if (e->dest == EXIT_BLOCK_PTR
+	      && last_stmt (bb) == NULL_TREE
 	      && TREE_CODE (last_stmt (bb)) == RETURN_EXPR
 	      && !single_pred_p (bb))
 	    {
diff -ruN gcc/real.c gcc/real.c
--- gcc/real.c	2007-10-22 06:54:31.692244904 +0000
+++ gcc/real.c	2007-10-22 07:20:13.140072521 +0000
@@ -2212,6 +2212,8 @@
   np2 = SIGNIFICAND_BITS - fmt->p * fmt->log2_b;
   memset (r->sig, -1, SIGSZ * sizeof (unsigned long));
   clear_significand_below (r, np2);
+  if (REAL_MODE_FORMAT_COMPOSITE_P (mode))
+      clear_significand_bit (r, SIGNIFICAND_BITS - fmt->pnan - 1);
 }
 
 /* Fills R with 2**N.  */
diff -ruN gcc/tree-dump.c gcc/tree-dump.c
--- gcc/tree-dump.c	2007-10-22 06:54:19.956987628 +0000
+++ gcc/tree-dump.c	2007-10-22 07:20:13.152075852 +0000
@@ -582,6 +582,10 @@
       dump_child ("args", TREE_OPERAND (t, 1));
       break;
 
+    case STATIC_CHAIN_EXPR:
+      dump_child ("func", TREE_OPERAND (t, 0));
+      break;
+
     case CONSTRUCTOR:
       {
 	unsigned HOST_WIDE_INT cnt;
diff -ruN gcc/tree-gimple.c gcc/tree-gimple.c
--- gcc/tree-gimple.c	2007-10-22 06:54:31.688243793 +0000
+++ gcc/tree-gimple.c	2007-10-22 07:20:13.200089176 +0000
@@ -72,6 +72,8 @@
     case VECTOR_CST:
     case OBJ_TYPE_REF:
     case ASSERT_EXPR:
+    case STATIC_CHAIN_EXPR: /* not sure if this is right...*/
+    case STATIC_CHAIN_DECL:
       return true;
 
     default:
@@ -143,7 +145,10 @@
 	  || TREE_CODE (t) == WITH_SIZE_EXPR
 	  /* These are complex lvalues, but don't have addresses, so they
 	     go here.  */
-	  || TREE_CODE (t) == BIT_FIELD_REF);
+	  || TREE_CODE (t) == BIT_FIELD_REF
+          /* This is an lvalue because it will be replaced with the real
+	     static chain decl. */
+	  || TREE_CODE (t) == STATIC_CHAIN_DECL);
 }
 
 /*  Return true if T is a GIMPLE condition.  */
diff -ruN gcc/tree-inline.c gcc/tree-inline.c
--- gcc/tree-inline.c	2007-10-22 06:54:53.982431834 +0000
+++ gcc/tree-inline.c	2007-10-22 07:20:13.208091397 +0000
@@ -602,10 +602,21 @@
      knows not to copy VAR_DECLs, etc., so this is safe.  */
   else
     {
+      if (! id->cloning_p && ! id->saving_p &&
+	  TREE_CODE (*tp) == MODIFY_EXPR &&
+	  TREE_OPERAND (*tp, 0) ==
+	  DECL_STRUCT_FUNCTION (fn)->static_chain_decl)
+	{
+	  /* Don't use special methods to initialize the static chain
+	     if expanding inline.  If this code could somehow be
+	     expanded in expand_start_function, it would not be
+	     necessary to deal with it here. */
+	  *tp = build_empty_stmt ();
+	}
       /* Here we handle trees that are not completely rewritten.
 	 First we detect some inlining-induced bogosities for
 	 discarding.  */
-      if (TREE_CODE (*tp) == MODIFY_EXPR
+      else if (TREE_CODE (*tp) == MODIFY_EXPR
 	  && TREE_OPERAND (*tp, 0) == TREE_OPERAND (*tp, 1)
 	  && (lang_hooks.tree_inlining.auto_var_in_fn_p
 	      (TREE_OPERAND (*tp, 0), fn)))
diff -ruN gcc/tree-nested.c gcc/tree-nested.c
--- gcc/tree-nested.c	2007-10-22 06:54:53.902409623 +0000
+++ gcc/tree-nested.c	2007-10-22 07:20:13.220094728 +0000
@@ -323,6 +323,7 @@
   if (!decl)
     {
       tree type;
+      enum tree_code code;
 
       type = get_frame_type (info->outer);
       type = build_pointer_type (type);
@@ -333,12 +334,17 @@
 	 Note also that it's represented as a parameter.  This is more
 	 close to the truth, since the initial value does come from 
 	 the caller.  */
-      decl = build_decl (PARM_DECL, create_tmp_var_name ("CHAIN"), type);
+      /* If the function has a custom static chain, a VAR_DECL is more
+	 appropriate. */
+      code = DECL_STRUCT_FUNCTION (info->context)->custom_static_chain ?
+	  VAR_DECL : PARM_DECL;
+      decl = build_decl (code, create_tmp_var_name ("CHAIN"), type);
       DECL_ARTIFICIAL (decl) = 1;
       DECL_IGNORED_P (decl) = 1;
       TREE_USED (decl) = 1;
       DECL_CONTEXT (decl) = info->context;
-      DECL_ARG_TYPE (decl) = type;
+      if (TREE_CODE (decl) == PARM_DECL)
+	  DECL_ARG_TYPE (decl) = type;
 
       /* Tell tree-inline.c that we never write to this variable, so
 	 it can copy-prop the replacement value immediately.  */
@@ -741,6 +747,8 @@
 
   if (info->context == target_context)
     {
+      /* might be doing something wrong to need the following line.. */
+      get_frame_type (info);
       x = build_addr (info->frame_decl, target_context);
     }
   else
@@ -1224,6 +1232,10 @@
       if (DECL_NO_STATIC_CHAIN (decl))
 	break;
 
+      /* Don't use a trampoline for a static reference. */
+      if (TREE_STATIC (t))
+	break;
+
       /* Lookup the immediate parent of the callee, as that's where
 	 we need to insert the trampoline.  */
       for (i = info; i->context != target_context; i = i->outer)
@@ -1288,6 +1300,14 @@
 	  = get_static_chain (info, target_context, &wi->tsi);
       break;
 
+    case STATIC_CHAIN_EXPR:
+      *tp = get_static_chain (info, TREE_OPERAND (t, 0), &wi->tsi);
+      break;
+
+    case STATIC_CHAIN_DECL:
+      *tp = get_chain_decl (info);
+      break;
+ 
     case RETURN_EXPR:
     case MODIFY_EXPR:
     case WITH_SIZE_EXPR:
@@ -1387,7 +1407,28 @@
       tree x = build (COMPONENT_REF, TREE_TYPE (root->chain_field),
 		      root->frame_decl, root->chain_field, NULL_TREE);
       x = build (MODIFY_EXPR, TREE_TYPE (x), x, get_chain_decl (root));
-      append_to_statement_list (x, &stmt_list);
+      /* If the function has a custom static chain, chain_field must
+	 be set after the static chain. */
+      if (DECL_STRUCT_FUNCTION (root->context)->custom_static_chain)
+	{
+	  /* Should use walk_function instead. */
+	  tree_stmt_iterator i =
+	      tsi_start ( BIND_EXPR_BODY (DECL_SAVED_TREE (context)));
+	  while (!tsi_end_p (i))
+	    {
+	      tree t = tsi_stmt (i);
+	      if (TREE_CODE (t) == MODIFY_EXPR &&
+		  TREE_OPERAND (t, 0) == root->chain_decl)
+		{
+		  tsi_link_after(& i, x, TSI_SAME_STMT);
+		  x = NULL_TREE;
+		  break;
+		}
+	    }
+	  gcc_assert(x == NULL_TREE);
+	}
+      else
+	append_to_statement_list (x, &stmt_list);
     }
 
   /* If trampolines were created, then we need to initialize them.  */
diff -ruN gcc/tree-pretty-print.c gcc/tree-pretty-print.c
--- gcc/tree-pretty-print.c	2007-10-22 06:54:19.984995402 +0000
+++ gcc/tree-pretty-print.c	2007-10-22 07:20:13.260105832 +0000
@@ -1004,6 +1004,16 @@
 	pp_string (buffer, " [tail call]");
       break;
 
+    case STATIC_CHAIN_EXPR:
+	pp_string (buffer, "<<static chain of ");
+	dump_generic_node (buffer, TREE_OPERAND (node, 0), spc, flags, false);
+	pp_string (buffer, ">>");
+      break;
+
+    case STATIC_CHAIN_DECL:
+       pp_string (buffer, "<<static chain decl>>");
+       break;
+	
     case WITH_CLEANUP_EXPR:
       NIY;
       break;
diff -ruN gcc/tree-sra.c gcc/tree-sra.c
--- gcc/tree-sra.c	2007-10-22 06:54:53.418275245 +0000
+++ gcc/tree-sra.c	2007-10-22 07:20:13.268108053 +0000
@@ -198,6 +198,8 @@
     case RECORD_TYPE:
       {
 	bool saw_one_field = false;
+	tree last_offset = size_zero_node;
+	tree cmp;
 
 	for (t = TYPE_FIELDS (type); t ; t = TREE_CHAIN (t))
 	  if (TREE_CODE (t) == FIELD_DECL)
@@ -207,6 +209,11 @@
 		  && (tree_low_cst (DECL_SIZE (t), 1)
 		      != TYPE_PRECISION (TREE_TYPE (t))))
 		goto fail;
+	      /* Reject aliased fields created by GDC for anonymous unions. */
+	      cmp = fold_binary_to_constant (LE_EXPR, boolean_type_node,
+		DECL_FIELD_OFFSET (t), last_offset);
+	      if (cmp == NULL_TREE || tree_expr_nonzero_p (cmp))
+		goto fail;
 
 	      saw_one_field = true;
 	    }
diff -ruN gcc/tree.def gcc/tree.def
--- gcc/tree.def	2007-10-22 06:54:52.758092003 +0000
+++ gcc/tree.def	2007-10-22 07:20:13.276110273 +0000
@@ -526,6 +526,13 @@
    Operand 2 is the static chain argument, or NULL.  */
 DEFTREECODE (CALL_EXPR, "call_expr", tcc_expression, 3)
 
+/* Operand 0 is the FUNC_DECL of the outer function for
+   which the static chain is to be computed. */
+DEFTREECODE (STATIC_CHAIN_EXPR, "static_chain_expr", tcc_expression, 1)
+    
+/* Represents a function's static chain.  It can be used as an lvalue. */
+DEFTREECODE (STATIC_CHAIN_DECL, "static_chain_decl", tcc_expression, 0)
+
 /* Specify a value to compute along with its corresponding cleanup.
    Operand 0 is the cleanup expression.
    The cleanup is executed by the first enclosing CLEANUP_POINT_EXPR,
diff -ruN gcc/varray.h gcc/varray.h
--- gcc/varray.h	2007-10-22 06:55:13.391819157 +0000
+++ gcc/varray.h	2007-10-22 07:20:13.284112494 +0000
@@ -62,6 +62,17 @@
   NUM_VARRAY_DATA
 };
 
+#ifndef __cplusplus 
+# define VARRAY_STRANGE_1 rtx
+# define VARRAY_STRANGE_2 rtvec
+# define VARRAY_STRANGE_3 tree
+#else
+# define VARRAY_STRANGE_1 rtx_
+# define VARRAY_STRANGE_2 rtvec_
+# define VARRAY_STRANGE_3 tree_
+#endif
+
+
 /* Union of various array types that are used.  */
 typedef union varray_data_tag GTY (()) {
   char			  GTY ((length ("%0.num_elements"),
@@ -91,11 +102,11 @@
   char			 *GTY ((length ("%0.num_elements"),
 				tag ("VARRAY_DATA_CPTR")))	cptr[1];
   rtx			  GTY ((length ("%0.num_elements"),
-				tag ("VARRAY_DATA_RTX")))	rtx[1];
+				tag ("VARRAY_DATA_RTX")))	VARRAY_STRANGE_1[1];
   rtvec			  GTY ((length ("%0.num_elements"),
-				tag ("VARRAY_DATA_RTVEC")))	rtvec[1];
+				tag ("VARRAY_DATA_RTVEC")))	VARRAY_STRANGE_2[1];
   tree			  GTY ((length ("%0.num_elements"),
-				tag ("VARRAY_DATA_TREE")))	tree[1];
+				tag ("VARRAY_DATA_TREE")))	VARRAY_STRANGE_3[1];
   struct bitmap_head_def *GTY ((length ("%0.num_elements"),
 				tag ("VARRAY_DATA_BITMAP")))	bitmap[1];
   struct reg_info_def	 *GTY ((length ("%0.num_elements"), skip,
